### PR TITLE
OCPBUGS-17528: Get kubeletCA from controller configuration

### PR DESCRIPTION
**Pull request by jrvaldes**
This change pulls the kubelet CA from the machine's controller configuration, given it is getting removed from the machine ignition files.

Refs:
- https://github.com/openshift/machine-config-operator/pull/3787

### CODE REVIEW

File: pkg/nodeconfig/nodeconfig.go [hunk 1]
```diff
@@ -346,25 +346,25 @@ func (nc *nodeConfig) createFilesFromIgnition() (map[string]string, error) {
...
-			}
-			contents, err := dataurl.DecodeString(*ignFile.Contents.Source)
-			if err != nil {
-				return nil, fmt.Errorf("could not decode %s: %w", ignFile.Node.Path, err)
+	// process kubelet-ca
+	filePathsToContents[windows.K8sDir+"\\"+KubeletClientCAFilename] = string(ign.GetKubeletCAData())
+	// process cloud config, if enabled
+	if _, ok := kubeletArgs[ignition.CloudConfigOption]; ok {
+		for _, ignFile := range ign.GetFiles() {
+			if ignFile.Node.Path == ignition.CloudConfigPath {
```
Comment (saifshaikh48) [inline pkg/nodeconfig/nodeconfig.go:355]: consider reducing nesting here as well
```
if ignFile.Node.Path != ignition.CloudConfigPath {
   continue
}
...
```

**Comment by mansikulkarni96** [comment]
LGTM

File: pkg/nodeconfig/nodeconfig.go [hunk 2]
```diff
@@ -346,25 +346,25 @@ func (nc *nodeConfig) createFilesFromIgnition() (map[string]string, error) {
...
-			if ignFile.Contents.Source == nil {
-				return nil, fmt.Errorf("could not process %s: File is empty", ignFile.Node.Path)
-			}
-			contents, err := dataurl.DecodeString(*ignFile.Contents.Source)
-			if err != nil {
-				return nil, fmt.Errorf("could not decode %s: %w", ignFile.Node.Path, err)
+	// process kubelet-ca
+	filePathsToContents[windows.K8sDir+"\\"+KubeletClientCAFilename] = string(ign.GetKubeletCAData())
+	// process cloud config, if enabled
+	if _, ok := kubeletArgs[ignition.CloudConfigOption]; ok {
```
[the remark this answers is not in the document]
Comment (jrvaldes) [reply pkg/nodeconfig/nodeconfig.go:353]: I did, and the above leaves the CloudConfig logic in the main execution path which is a bit out of context as it is specific/only relevant when the CloudConfig option is present.

File: pkg/nodeconfig/nodeconfig.go (hunk 1 above)
Comment (jrvaldes) [reply pkg/nodeconfig/nodeconfig.go:355]: Addressed.

**Comment by jrvaldes** [comment]
Blocked by:

- https://github.com/openshift/windows-machine-config-operator/pull/1728

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

File: pkg/nodeconfig/nodeconfig.go [hunk 3]
```diff
@@ -346,25 +346,26 @@ func (nc *nodeConfig) createFilesFromIgnition() (map[string]string, error) {
...
 			contents, err := dataurl.DecodeString(*ignFile.Contents.Source)
 			if err != nil {
-				return nil, fmt.Errorf("could not decode %s: %w", ignFile.Node.Path, err)
+				return nil, fmt.Errorf("could not decode %s: %w", ignition.CloudConfigPath, err)
 			}
-			fileName := filepath.Base(ignFile.Node.Path)
+			fileName := filepath.Base(ignition.CloudConfigPath)
 			filePathsToContents[windows.K8sDir+"\\"+fileName] = string(contents.Data)
+
+			break
```
Comment (sebsoto) [inline pkg/nodeconfig/nodeconfig.go:368]: This seems like an unnecessary refactor in a bug fix.

Comment (jrvaldes) [reply pkg/nodeconfig/nodeconfig.go]: @sebsoto  [Addressed](https://github.com/openshift/windows-machine-config-operator/compare/49047ec9fe7ee9ffdb78cbfae8d37835435eb3b3..e588e5c73f1d74403a52332e27b677bfe58de4e3)

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

File: pkg/nodeconfig/nodeconfig.go [hunk 4]
```diff
@@ -346,13 +346,17 @@ func (nc *nodeConfig) createFilesFromIgnition() (map[string]string, error) {
...
-	filesToTransfer := map[string]struct{}{
-		ignition.KubeletCACertPath: {},
-	}
+	filesToTransfer := map[string]struct{}{}
 	if _, ok := kubeletArgs[ignition.CloudConfigOption]; ok {
 		filesToTransfer[ignition.CloudConfigPath] = struct{}{}
 	}
 	filePathsToContents := make(map[string]string)
+	// process kubelet-ca
+	filePathsToContents[windows.K8sDir+"\\"+KubeletClientCAFilename] = string(ign.GetKubeletCAData())
```
Comment (sebsoto) [inline pkg/nodeconfig/nodeconfig.go:355]: if `ign.GetKubeletCAData()` returns an empty string, the file will be created with no contents. Is that the expected behavior? Or should this be to not create the file at all?

if so you should do this:
```
if caData := string(ign.GetKubeletCAData()); caData != "" {
     	filePathsToContents[windows.K8sDir+"\\"+KubeletClientCAFilename] = string(caData)
}
```

**Review by sebsoto — APPROVED** [review]

File: pkg/nodeconfig/nodeconfig.go (hunk 4 above)
Comment (jrvaldes) [reply pkg/nodeconfig/nodeconfig.go:355]: I don't feel this is a valid case in the execution flow, given the kubeletCA is [required](https://github.com/openshift/windows-machine-config-operator/pull/1725/files#diff-5f8bc59c6ca4160f53648587fbed538adbd305ad0dfb837bd3124f2ee84fde9bR78).

**Comment by alinaryan** [comment]
Thanks for working on this, 
/lgtm

File: pkg/nodeconfig/nodeconfig.go (hunk 4 above)
Comment (aravindhp) [reply pkg/nodeconfig/nodeconfig.go:355]: This implies, if for whatever reason, the kubelet CA is empty, we are going to silently fail. How are you handling that?

File: pkg/ignition/ignition.go [hunk 5]
```diff
@@ -60,9 +61,33 @@ func New(c client.Client) (*Ignition, error) {
...
+	for _, item := range ccList.Items {
+		if item.Spec.KubeAPIServerServingCAData != nil {
+			log.V(1).Info("processing kubelet-ca", "ControllerConfig", item.Name)
+			kubeletCAData = item.Spec.KubeAPIServerServingCAData
+			break
+		}
+	}
+	if len(kubeletCAData) == 0 {
+		return nil, fmt.Errorf("cannot find kubelet-ca")
+	}
```
Comment (aravindhp) [inline pkg/ignition/ignition.go:79]: Not asking for a change but this sort of changes should be in a helper function.

File: pkg/nodeconfig/nodeconfig.go (hunk 4 above)
Comment (jrvaldes) [reply pkg/nodeconfig/nodeconfig.go:355]: Below the link behind ["required"](https://github.com/openshift/windows-machine-config-operator/pull/1725/files#diff-5f8bc59c6ca4160f53648587fbed538adbd305ad0dfb837bd3124f2ee84fde9bR78) in the above comment where `ignition.New(nc.client)` returns an error. 

https://github.com/openshift/windows-machine-config-operator/pull/1725/files#diff-5f8bc59c6ca4160f53648587fbed538adbd305ad0dfb837bd3124f2ee84fde9bR78

File: pkg/ignition/ignition.go (hunk 5 above)
Comment (jrvaldes) [reply pkg/ignition/ignition.go:79]: Agree. I'll take it for a follow-up PR.

File: pkg/nodeconfig/nodeconfig.go (hunk 4 above)
Comment (aravindhp) [reply pkg/nodeconfig/nodeconfig.go:355]: OK, that satisfies my concern

**Comment by cdoern** [comment]
Does this need to get rebased on top of the ignition PR to pass the tests?

**Comment by jrvaldes** [comment]
> Does this need to get rebased on top of the ignition PR to pass the tests?

[Done](https://github.com/openshift/windows-machine-config-operator/compare/e588e5c73f1d74403a52332e27b677bfe58de4e3..db7a734cd1ae1ef4294d4e57a58b9555fbfdf389)

**Comment by aravindhp** [comment]
Explicit rebasing need not be done. Prow takes care of that for us when testing.

**Review by aravindhp — APPROVED** [review]
/lgtm

**Comment by jrvaldes** [comment]
/test vsphere-proxy-e2e-operator 

timed out handling build

**Comment by aravindhp** [comment]
/test vsphere-proxy-e2e-operator 

Previous [run](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1725/pull-ci-openshift-windows-machine-config-operator-master-vsphere-proxy-e2e-operator/1692170321194389504) failed in cluster install
```
level=info msg=Obtaining RHCOS image file from 'https://rhcos.mirror.openshift.com/art/storage/prod/streams/4.14-9.2/builds/414.92.202308032115-0/x86_64/rhcos-414.92.202308032115-0-vmware.x86_64.ova?sha256=dcc75d586ad2c6e7c0eb68fa738a3fb40100f2522840b26cdf105760c2c42002'
level=error msg=File sha256 checksum is invalid.
level=fatal msg=failed to fetch Terraform Variables: failed to generate asset "Terraform Variables": failed to get vsphere Terraform variables: failed to use cached vsphere image: Checksum mismatch for /tmp/.cache/openshift-installer/image_cache/rhcos-414.92.202308032115-0-vmware.x86_64.ova; expected=dcc75d586ad2c6e7c0eb68fa738a3fb40100f2522840b26cdf105760c2c42002 found=fd0c404e3fe2322a22f1907d19a5da387242fd138e06d3cd8ac612c020539f22
Installer exit with code 1
```

**Comment by aravindhp** [comment]
/override ci/prow/gcp-e2e-operator

The job is prema failing due to [OCPBUGS-17731](https://issues.redhat.com/browse/OCPBUGS-17731)

**Pull request merged**